### PR TITLE
Performance optimize invokables (bug fix also)

### PR DESCRIPTION
**Pull request by fhein**
Current master and develop branches provide a special treatment for invokables within sm configuration phase. This treatment requires to loop through the invokables array and perform actions and create an InvokableFactory entry for each Invokable and an alias if the name of the invokable is different from the invokable class name. This is a major bottleneck for configuration speed. This treatment alone takes about 30% of overall configuration time.

Here is the benchmark comparing master to this PR.

    benchmark: FetchNewServiceManagerBench
	+----------------------------------+-------------------+------------------+--------+
	| subject                          | suite:master:mean | suite:PR232:mean | factor |
	+----------------------------------+-------------------+------------------+--------+
	| benchFetchServiceManagerCreation | 872.000µs         | 218.000µs        | 4x     |
	+----------------------------------+-------------------+------------------+--------+

The replacement of an invokable by a factory and an alias can also be considered as a bug. Current service resolution precedence is:

1. Services
2. Aliases
3. Delegators
4. Factories
5. Abstract Factories

As @Ocramius said, there are use cases, where users want to provide a factory for an invokable class, because the particular application wants to apply configuration to the invokable object.

By defining an alias for the invokable class, this factory gets disabled efffectively.

This PR changes the treatment of invokables. They are stored in an $invokables array like the other items.
Now invokables get resolved to an InvokableFactory in getFactory now. The new resolution precedence is

1. Services
2. Aliases
3. Delegators
4. Factories
5. Invokables
6. Abstract Factories

I tried to discuss the problems of alias resolution precedence in #222. I do not know if there are discussions around that topic actually. At least nobody wanted to share his/her thoughts since I published that article weeks ago. Trouble with aliases could get eliminated by reducing them to aliases. That would mean to change the resolution precedence to 

1. Services
2. Delegators
3. Factories
4. Invokables
5. Abstract Factories
6. Aliases

### CODE REVIEW

File: benchmarks/BenchAsset/DelegatorFactoryFoo.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
+<?php
+/**
+ * @link      http://github.com/zendframework/zend-servicemanager for the canonical source repository
+ * @copyright Copyright (c) 2016 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   http://framework.zend.com/license/new-bsd New BSD License
```
Comment (weierophinney) [inline benchmarks/BenchAsset/DelegatorFactoryFoo.php]: On this and other file-level docblocks for new files introduced by this patch, please do the following:

- Use https scheme for all URLs
- Date is 2018

File: benchmarks/SetNewServicesBench.php [hunk 2]
```diff
@@ -80,11 +83,58 @@ public function benchSetAlias()
...
         // @todo @link https://github.com/phpbench/phpbench/issues/304
         $sm = clone $this->sm;
 
         $sm->setAlias('recursiveFactoryAlias1', 'factory1');
     }
+
+    public function benchSetInvokableClass()
+    {
+
+        // @todo @link https://github.com/phpbench/phpbench/issues/304
```
Comment (weierophinney) [inline benchmarks/SetNewServicesBench.php]: On this and the following new methods:

- Move the `@todo` annotation to a method-level docblock
- Remove the blank line at the top of the method

File: src/ServiceManager.php [hunk 3]
```diff
@@ -413,11 +404,11 @@ public function configure(array $config)
...
     {
-        if (isset($this->services[$alias]) && ! $this->allowOverride) {
-            throw ContainerModificationsNotAllowedException::fromExistingService($alias);
+        if (! isset($this->services[$alias]) || $this->allowOverride) {
+            $this->mapAliasToTarget($alias, $target);
+            return;
         }
-
-        $this->mapAliasToTarget($alias, $target);
+        throw ContainerModificationsNotAllowedException::fromExistingService($alias);
```
Comment (weierophinney) [inline src/ServiceManager.php]: Please revert this change; exceptions should be thrown within conditionals for internal consistency and with other components.

File: src/ServiceManager.php [hunk 4]
```diff
@@ -431,11 +422,11 @@ public function setAlias($alias, $target)
...
     {
-        if (isset($this->services[$name]) && ! $this->allowOverride) {
-            throw ContainerModificationsNotAllowedException::fromExistingService($name);
+        if (! isset($this->services[$name]) || $this->allowOverride) {
+            $this->invokables[$name] = $class ?? $name;
+            return;
         }
-
-        $this->createAliasesAndFactoriesForInvokables([$name => $class ?? $name]);
+        throw ContainerModificationsNotAllowedException::fromExistingService($name);
```
Comment (weierophinney) [inline src/ServiceManager.php]: Invert the conditional and throw from the conditional, please.

File: src/ServiceManager.php [hunk 5]
```diff
@@ -449,23 +440,27 @@ public function setInvokableClass($name, $class = null)
...
-        if (isset($this->services[$name]) && ! $this->allowOverride) {
-            throw ContainerModificationsNotAllowedException::fromExistingService($name);
+        if (! isset($this->services[$name]) || $this->allowOverride) {
+            $this->factories[$name] = $factory;
+            return;
         }
-
-        $this->factories[$name] = $factory;
+        throw ContainerModificationsNotAllowedException::fromExistingService($name);
     }
```
Comment (weierophinney) [inline src/ServiceManager.php]: Please revert this change; exceptions should be thrown within conditionals for internal consistency and with other components.

File: src/ServiceManager.php [hunk 6]
```diff
@@ -449,23 +440,27 @@ public function setInvokableClass($name, $class = null)
...
      *
-     * @param string $name Service name to map
-     * @param null|string $class Class to which to map; if not provided, $name
-     *     will be used for the mapping.
+     * @param string $name Service name
+     * @param string|callable|Factory\DelegatorFactoryInterface $factory Delegator
+     *     factory to assign.
      */
-    public function mapLazyService($name, $class = null)
+    public function addDelegator($name, $factory)
```
Comment (weierophinney) [inline src/ServiceManager.php]: Why did you switch the positions of these two methods? Please revert, and also revert the changes within each.

File: src/ServiceManager.php [hunk 7]
```diff
@@ -511,10 +511,11 @@ public function addInitializer($initializer)
...
     {
-        if (isset($this->services[$name]) && ! $this->allowOverride) {
-            throw ContainerModificationsNotAllowedException::fromExistingService($name);
+        if (! isset($this->services[$name]) || $this->allowOverride) {
+            $this->services[$name] = $service;
+            return;
         }
-        $this->services[$name] = $service;
+        throw ContainerModificationsNotAllowedException::fromExistingService($name);
     }
```
Comment (weierophinney) [inline src/ServiceManager.php]: Please revert this change; exceptions should be thrown within conditionals for internal consistency and with other components.

File: src/ServiceManager.php [hunk 8]
```diff
@@ -527,11 +528,11 @@ public function setService($name, $service)
...
-        if (isset($this->services[$name]) && ! $this->allowOverride) {
-            throw ContainerModificationsNotAllowedException::fromExistingService($name);
+        if (! isset($this->services[$name]) || $this->allowOverride) {
+            $this->shared[$name] = (bool) $flag;
+            return;
         }
-
-        $this->shared[$name] = (bool) $flag;
+        throw ContainerModificationsNotAllowedException::fromExistingService($name);
     }
```
Comment (weierophinney) [inline src/ServiceManager.php]: Please revert this change; exceptions should be thrown within conditionals for internal consistency and with other components.

File: src/ServiceManager.php [hunk 9]
```diff
@@ -868,6 +858,12 @@ private function mapAliasToTarget($alias, $target)
      *
      * This function maps $this->aliases in place.
      *
+     * This algorithm is an adaptated version of Tarjans Strongly
+     * Connected Components. Instead of returning the strongly
+     * connected components (i.e. cycles in our case), we throw.
+     * If nodes are not strongly connected (i.e. resolvable in
+     * our case), they get resolved.
+     *
```
Comment (weierophinney) [inline src/ServiceManager.php]: The changes in this method appear to be from #230; please rebase, so that this patch only contains the changes you propose.

File: src/ServiceManager.php [hunk 10]
```diff
@@ -899,25 +915,26 @@ private function mapAliasesToTargets()
      * Instantiate abstract factories in order to avoid checks during service construction.
      *
      * @param string[]|Factory\AbstractFactoryInterface[] $abstractFactories
-     *
-     * @return void
      */
-    private function resolveAbstractFactoryInstance($abstractFactory)
+    private function resolveAbstractFactories($abstractFactories)
```
Comment (weierophinney) [inline src/ServiceManager.php]: The changes in this method appear to be from a previously merged commit; please rebase, so that this patch only contains the changes you propose.

File: test/CommonServiceLocatorBehaviorsTrait.php [hunk 11]
```diff
@@ -855,4 +854,28 @@ public function testCrashesOnCyclicAliases()
...
+            ],
+            'aliases' => [
+                'alias1' => 'alias2',
+                'alias2' => 'alias3',
+                'alias3' => stdClass::class,
+            ],
+        ]);
+        $this->assertSame($sm->get('alias1'), $sm->get('alias2'));
+        $this->assertSame($sm->get(stdClass::class), $sm->get('alias1'));
+    }
```
Comment (weierophinney) [inline test/CommonServiceLocatorBehaviorsTrait.php]: These two changes are from #230; please rebase.

File: src/ServiceManager.php (hunk 3 above)
Comment (fhein) [reply src/ServiceManager.php]: My idea behind that is, that by default a configuration should get assumed to be error free (it finally is after all errors got fixed). Assumed the configuration is error free, then  the `||` clause (with appropriate ordering,  which I am not quite sure yet what the best is) can pass successfully (`true`) after evaluation of just the first expression of the `||` clause at a high probability. 

I implemented it in a way, that one and worst case two evaluations are necessary to determine, that operation can continue, assuming that to continue occurs much more often than failure in production scenarios.

Inverting the condition means, that one maybe two evalutions get necessary to determine not to throw. The assumption here is pessimistic. Because working configs will be error free, I optimized assuming that case. It does not matter, how long it takes to decide to throw.

What do you think? Did I miss to consider something important?  

I hope my English is good enough to express the idea.

Anyway, this is fine tuning.

Comment (weierophinney) [reply src/ServiceManager.php]: There are two points I was making:

- Consistency. Throughout the framework, we generally (as in greater than 99% of cases) _throw from a conditional_. The method body is where the work of the method is done. In these cases, you're setting an alias, setting a factory, etc. That's the work of the methods. The conditional should be used to flag _exceptional cases that prevent that work from happening_.

- Common use case. The primary use case for zend-servicemanager is to _configure once_. These setters are not the primary path. As such, having them readable and consistent with the code base is more important than micro-optimizations from conditionals.

If you are worried about the conditional performance, due to needing to evaluate two conditions, put the one most likely to be `false` first — which is how these were done (i.e., checking to see if the service/alias/factory/etc. has already been set). If the service has not been previously set, only the first part of the condition will be evaluated, and we'll then move on to the next statement.

On top of that, boolean checks (`&& ! $this->allowOverride`) are quite fast, and inverting the method workflow for this is truly a micro-optimization at the expense of consistency and readability.

Comment (fhein) [reply src/ServiceManager.php]: Ok.

File: src/ServiceManager.php (hunk 9 above)
Comment (fhein) [reply src/ServiceManager.php]: I suggest I rebase to #231, if it be accepted. With #231 I did much you requested.

File: test/CommonServiceLocatorBehaviorsTrait.php (hunk 11 above)
Comment (fhein) [reply test/CommonServiceLocatorBehaviorsTrait.php]: I suggest I rebase to #231, if that can get accepted. With #231 I did much you requested.

**Comment by fhein** [comment]
Closed because this primarily a bug fix for invokable handling. See #242.
